src/runner_metrics.py

```diff
@@ -0,0 +1,198 @@
+    """
+
+    timestamp: NonNegativeFloat
+    status: str
+
+
+class RunnerMetrics(BaseModel):
+    """Metrics for a runner.
+
+    Args:
+        installed_timestamp: The UNIX time stamp of the time at which the runner was installed.
+        pre_job: The metrics for the pre-job phase.
+        post_job: The metrics for the post-job phase.
+    """
+
+    installed_timestamp: NonNegativeFloat
+    pre_job: PreJobMetrics
+    post_job: Optional[PostJobMetrics]
+
+
+def _inspect_file_sizes(fs: shared_fs.SharedFilesystem) -> None:
+    """Inspect the file sizes of the shared filesystem.
+
+    Args:
+        fs: The path to the shared filesystem for a specific runner.
+
+    Raises:
+        FileSizeTooLargeError: If one of the files is too large.
+    """
+    files = [
+        fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(POST_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME),
+    ]
+    for file in files:
+        if file.exists() and file.stat().st_size > FILE_SIZE_BYTES_LIMIT:
+            raise MetricFileSizeTooLargeError(
+                f"File {file} is too large. The limit is {FILE_SIZE_BYTES_LIMIT} bytes."
+            )
+
+
+def _extract_metrics_from_fs(fs: shared_fs.SharedFilesystem) -> Optional[RunnerMetrics]:
+    """Extract metrics from a shared filesystem.
+
+    Args:
+        fs: The path to the shared filesystem for a specific runner.
+
+    Returns:
+        The extracted metrics if at least the pre-job metrics are present.
+
+    Raises:
+        JSONDecodeError: If the pre-job or post-job metrics file is not valid JSON.
+        pydantic.ValidationError: If one of the files is not valid.
+        FileSizeTooLargeError: If one of the files is too large.
+        FileNotFoundError: If installed_timestamp is not found.
+    """
+    _inspect_file_sizes(fs)
+
+    installed_timestamp = fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME).read_text()
+    logger.debug("Runner %s installed at %s", fs.runner_name, installed_timestamp)
+    try:
+        pre_job_metrics = json.loads(fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME).read_text())
+        logger.debug("Pre-job metrics for runner %s: %s", fs.runner_name, pre_job_metrics)
+    except FileNotFoundError:
+        logger.warning("%s not found for runner %s.", PRE_JOB_METRICS_FILE_NAME, fs.runner_name)
+        return None
+
+    try:
+        post_job_metrics = json.loads(fs.path.joinpath(POST_JOB_METRICS_FILE_NAME).read_text())
+        logger.debug("Post-job metrics for runner %s: %s", fs.runner_name, post_job_metrics)
+    except FileNotFoundError:
+        logger.warning("%s not found for runner %s", POST_JOB_METRICS_FILE_NAME, fs.runner_name)
+        post_job_metrics = None
+
+    return RunnerMetrics(
+        installed_timestamp=installed_timestamp,
+        pre_job=PreJobMetrics(**pre_job_metrics),
+        post_job=PostJobMetrics(**post_job_metrics) if post_job_metrics else None,
```

Comment:
What about if the `pre_job_metrics` or `post_job_metrics` are valid JSON but not a dictionary?